roxie/udplib/udptrr.cpp

```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
             flow_socket->set_receive_buffer_size(udpFlowSocketsSize);
             size32_t actualSize = flow_socket->get_receive_buffer_size();
             DBGLOG("UdpReceiver: receive_receive_flow created port=%d sockbuffsize=%d actual %d", flow_port, udpFlowSocketsSize, actualSize);
+            Thread::start();
         }
-        
-        ~receive_receive_flow() 
+
+        void doFlowRequest(const UdpRequestToSendMsg &msg)
         {
-            running = false;
-            if (flow_socket) 
-                flow_socket->close();
-            join();
+            flowRequestsReceived++;
+            if (udpTraceLevel > 5 || udpTraceFlow)
+            {
+                StringBuffer ipStr;
+                DBGLOG("UdpReceiver: received %s msg flowSeq %" SEQF "u sendSeq %" SEQF "u from node=%s", flowType::name(msg.cmd), msg.flowSeq, msg.sendSeq, msg.sourceNode.getTraceText(ipStr).str());
+            }
+            UdpSenderEntry *sender = &parent.sendersTable[msg.sourceNode];
+
+            //If the flowSeq is lower than the previous flowSeq there could be two causes:
+            //a) The sender has restarted
+            //b) Messages have been received out of order (e.g. equest_to_send_more after a request_to_send?)
+            //=> ignore the flow if it was the previous seq, and could not come from the server restarting
+            unsigned flowSeq = msg.flowSeq;
+            const unsigned startupFlowSeqs = 2; // flow seq below this could be from a restarting server
+            if ((flowSeq + 1 == sender->flowSeq) && (flowSeq >= startupFlowSeqs))
```

Comment:
Can you never be more than one ahead?

Reply:
I think so, but needs to be checked, and rethought if completed flow messages are sent on the flow thread.